cmd/juju/controller/configcommand.go

```diff
@@ -0,0 +1,232 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package controller
+
+import (
+	"bytes"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+	"github.com/juju/utils/set"
+
+	apicontroller "github.com/juju/juju/api/controller"
+	"github.com/juju/juju/cmd/juju/common"
+	"github.com/juju/juju/cmd/modelcmd"
+	"github.com/juju/juju/cmd/output"
+	"github.com/juju/juju/controller"
+)
+
+// NewConfigCommand returns a new command that can retrieve or update
+// controller configuration.
+func NewConfigCommand() cmd.Command {
+	return modelcmd.WrapController(&configCommand{})
+}
+
+// configCommand is able to output either the entire environment or
+// the requested value in a format of the user's choosing.
+type configCommand struct {
+	modelcmd.ControllerCommandBase
+	api controllerAPI
+	out cmd.Output
+
+	action     func(controllerAPI, *cmd.Context) error // The action we want to perform, set in cmd.Init.
+	key        string                                  // One config key to read.
+	setOptions common.ConfigFlag                       // Config values to set.
+}
+
+const configCommandHelpDoc = `
+By default, all configuration (keys and values) for the controller are
+displayed if a key is not specified. Supplying one key name returns
+only the value for that key.
+
+Supplying key=value will set the supplied key to the supplied value;
+this can be repeated for multiple keys. You can also specify a yaml
+file containing key values. Not all keys can be updated after
+bootstrap time.
+
+Available keys and values can be found here:
+https://jujucharms.com/docs/stable/controllers-config
+
+Examples:
+
+    juju controller-config
+    juju controller-config api-port
+    juju controller-config -c mycontroller
+    juju controller-config auditing-enabled=true audit-log-max-backups=5
+    juju controller-config auditing-enabled=true path/to/file.yaml
+    juju controller-config path/to/file.yaml
+
+See also:
+    controllers
+    model-config
+    show-cloud
+`
+
+func (c *configCommand) Info() *cmd.Info {
+	return &cmd.Info{
+		Name:    "controller-config",
+		Args:    "[<attribute key>[=<value>] ...]",
+		Purpose: "Displays or sets configuration settings for a controller.",
+		Doc:     strings.TrimSpace(configCommandHelpDoc),
+	}
+}
+
+func (c *configCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.ControllerCommandBase.SetFlags(f)
+	c.out.AddFlags(f, "tabular", map[string]cmd.Formatter{
+		"json":    cmd.FormatJson,
+		"tabular": formatConfigTabular,
+		"yaml":    cmd.FormatYaml,
+	})
+}
+
+func (c *configCommand) Init(args []string) error {
```

Comment:
I know, pedantic, but comments for these ideally

Reply:
I'll add (even though these aren't new! :)